### DISCUSSION TITLE
Computing hop count in TcpSessionApp

### DISCUSSION THREAD

**Discussion by phdavis1027** [inet-framework/inet#957] [General]
How do I do basically [this](https://groups.google.com/g/omnetpp/c/iGT4fbAuYYk/m/dKAyhwwlEAAJ) in modern versions of inet, except in TcpSessionApp? My first idea was to extend TcpSessionApp and override `socketDataArrived` but the `Packet` class seems to be kind of opaque. Really, I would just like to be pointed in the direction of utilities for peeking inside of packets.

**Comment by rhornig** [comment]
`@statistics` can use packet filters to provide statistics about specific packet attributes. Check the `ResultFilters.h | .cc` files. The `messageAge` and `dataAge` filters do something very similar, you want to achieve (let's call it `dataHopCount`. Then, you have to add a statistic to the TcpSessionApp that is similar to the `endToEndDelay` statistic. Just it should use your `dataHopCount` filter. The data you need to access in the result filter implementation is available in the `HopLimitInd` indicator tag added during the IPv4 decapsilation.

**Reply by phdavis1027** [reply]
This is quite useful! I'm working on following your suggestion. Just out of curiosity, if I get this working, would there be interest in a PR to include hopCount as a statistic in TcpSessionApp by default?

**Reply by phdavis1027** [reply]
Quick followup: since I will be reusing the modules that collect  HopCount across multiple simulations, I tried to break them out into a `common` dir which I passed to `opp_makemake` with the -I flag. If I build and run with this structure, the generated makefile doesn't build a HopCounterFilter.o . However, if I build and run with HopCounterFilter.{c, h} in the project directory, it compiles and runs just fine. Is there are a way to tell the makefile about `common`?

**Reply by rhornig** [reply]
It depends, where that `common` dir is located. If it is under the <inet>/src folder, then the IDE will scan and include all `.cc` files automatically. From command line, you must re-execute `opp_makemake` to re-create a makefile that includes your new files.

- If the `common` dir is outside of the `src` folder and you want to share the code between **projects** this pretty much creates a shared library with your common code. This might be pretty tricky to set up correctly especially if you want to build it from both the IDE and the command line.
- If you just need a solution that works in house, I would just symlink the common directory somewhere inside the `src` folder at the appropriate place (in all directories where you use).
- or a variation of this: put your common code into a separate git repo and create a git submodule from it inside the `src` tree. This is a bit more involved, that a simple simlink, but it would work for other people too...

So, if you just want to use the project yourself, go with symlinking.

As for the PR: I'm not sure how useful would be this feature for the general public. I mean, at the end of the day, you will get an `average` hop count for a tcp stream (or for all TCP streams in that app?). But that could also be calculated at the network layer. It all depends on how complicated the code. If it's 3-4 well placed lines, we can look into this, but on long term we have to maintain that code, so anything longer is not necessary justified.
